README.rst
💎 Moldavite
------------

.. note::

  Moldavite is a forest green, olive green or blue greenish vitreous silica
  projectile rock formed by a meteorite impact probably in southern Germany.

Source: `Wikipedia <https://en.wikipedia.org/wiki/Moldavite>`__

Application
===========

This application creates functionality similar to `mybinder.org
<https://mybinder.org/>`__ service. A user supplies Git URL repository, path to
`JupyterBook <https://jupyterbook.org/>`__ and other information (such as Git
branch, lifetime of the deployed JupyterBook instance, ...) and the service
builds a JupyterBook that is hosted on OpenShift for the specified lifetime.

Requirements for deployment
===========================

* OpenShift cluster
* Argo Workflows
* Thoth's `cleanup-job <https://github.com/thoth-station/cleanup-job>`__ - used for
  cleaning resources in the cluster
* privileged containers for running ``podman``/``buildah`` to create JupyterBook


See `example repo for example input <https://github.com/fridex/moldavite-example>`__.

Running the application locally
===============================

The API server can be run locally without being deployed to the cluster. You
need to log in using ``oc`` and be active in the desired namespace. The API
server detects that it is running locally and talks to the Kubernetes API under
the hood, as if the API server would be run in a cluster:

.. code:: console

  pipenv run gunicorn moldavite.entrypoint:app

Check ``.env.template`` for setting up required environment variables and
`thoth-common <https://github.com/thoth-station/common>`__ for logging setup.


CHANGELOG.md
# Changelog for Moldavite API

## Release 0.0.1 (2021-08-09T11:48:08)
### Features
* Add mypy.ini file
* Fix typing errors
* Initial project import

## [0.0.0] - 2021-July-19 - fridex

### Added

* initial project import


moldavite/entrypoint.py
#!/usr/bin/env python3
# Moldavite
# Copyright(C) 2021 The Moldavite Authors.
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Core API configuration and entrypoint."""

import os
import sys
import logging
from typing import Any
from typing import Dict
from typing import Tuple

import connexion

from flask import redirect, jsonify, request
from flask_script import Manager
from flask_cors import CORS
from prometheus_flask_exporter import PrometheusMetrics
from werkzeug import Response

from thoth.common import datetime2datetime_str
from thoth.common import init_logging
from thoth.common import __version__ as __common__version__

from datetime import datetime

from moldavite import __version__
from moldavite.configuration import Configuration


# Configure global application logging using Thoth's init_logging.
init_logging(logging_env_var_start="THOTH_MOLDAVITE_LOG_")

_THOTH_API_HTTPS = bool(int(os.getenv("THOTH_API_HTTPS", 1)))
_LOGGER = logging.getLogger("moldavite")
_LOGGER.setLevel(
    logging.DEBUG if bool(int(os.getenv("THOTH_MOLDAVITE_DEBUG", 0))) else logging.INFO
)

__service_version__ = f"{__version__}+common.{__common__version__}"

_LOGGER.info("This is Moldavite API v%s", __service_version__)

# Expose for uWSGI.
app = connexion.App(__name__)
application = app.app

app.add_api(Configuration.SWAGGER_YAML_PATH)
metrics = PrometheusMetrics(application)
manager = Manager(application)

# Needed for session.
application.secret_key = Configuration.APP_SECRET_KEY

# static information as metric
metrics.info("moldavite_api_info", "Moldavite API info", version=__service_version__)

# Add Cross Origin Request Policy to all
CORS(app.app)


@app.route("/")  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def base_url() -> Response:
    """Redirect to UI by default."""
    # https://github.com/pallets/flask/issues/773
    request.environ["wsgi.url_scheme"] = "https" if _THOTH_API_HTTPS else "http"
    return redirect("api/v1/ui")


@app.route("/api/v1")  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def api_v1() -> Any:
    """Provide a listing of all available endpoints."""
    paths = []

    for rule in application.url_map.iter_rules():
        rule = str(rule)
        if rule.startswith("/api/v1"):
            paths.append(rule)

    return jsonify({"paths": paths})


def _healthiness() -> Tuple[Dict[str, str], int, Dict[str, str]]:
    return (
        jsonify({"status": "ready", "version": __service_version__}),
        200,
        {"ContentType": "application/json"},
    )


@app.route("/readiness")  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def api_readiness() -> Tuple[
    Dict[
        str,
        str,
    ],
    int,
    Dict[str, str],
]:
    """Report readiness for OpenShift readiness probe."""
    return _healthiness()


@app.route("/liveness")  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def api_liveness() -> Tuple[Dict[str, str], int, Dict[str, str]]:
    """Report liveness for OpenShift readiness probe."""
    return _healthiness()


@application.errorhandler(404)  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def page_not_found(exc: Exception) -> Tuple[Dict[str, str], int]:
    """Adjust 404 page to be consistent with errors reported back from API."""
    # Flask has a nice error message - reuse it.
    return jsonify({"error": str(exc)}), 404


@application.errorhandler(500)  # type: ignore[misc]
@metrics.do_not_track()  # type: ignore[misc]
def internal_server_error(exc: Exception) -> Tuple[Dict[str, str], int]:
    """Adjust 500 page to be consistent with errors reported back from API."""
    # Provide some additional information so we can easily find exceptions in logs (time and exception type).
    # Later we should remove exception type (for security reasons).
    return (
        jsonify(
            {
                "error": "Internal server error occurred, please contact administrator with provided details.",
                "details": {
                    "type": exc.__class__.__name__,
                    "datetime": datetime2datetime_str(datetime.utcnow()),
                },
            }
        ),
        500,
    )


@application.after_request  # type: ignore[misc]
def apply_headers(response: Response) -> Response:
    """Add headers to each response."""
    response.headers["X-Moldavite-Version"] = __service_version__
    return response


if __name__ == "__main__":
    app.run()

    sys.exit(1)


moldavite/configuration.py
#!/usr/bin/env python3
# Moldavite
# Copyright(C) 2021 The Moldavite Authors.
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Configuration of API service."""

import logging
import os

_LOGGER = logging.getLogger(__name__)


class Configuration:
    """Configuration of API service."""

    APP_SECRET_KEY = os.environ["THOTH_MOLDAVITE_API_APP_SECRET_KEY"]
    SWAGGER_YAML_PATH = os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "../openapi/openapi.yaml"
    )
    BUILD_NAMESPACE = os.environ["THOTH_MOLDAVITE_BUILD_NAMESPACE"]
    INFRA_NAMESPACE = os.environ["THOTH_MOLDAVITE_INFRA_NAMESPACE"]

    OPENAPI_PORT = 8080


moldavite/api_v1.py
#!/usr/bin/env python3
# Moldavite
# Copyright(C) 2021 The Moldavite Authors.
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Implementation of API v1."""

import os
import logging
from typing import Any
from typing import Dict
from typing import Tuple
from typing import Union

import datetime
from dateutil.parser import parse as datetime_parser
from thoth.common.exceptions import NotFoundExceptionError
from thoth.common import OpenShift

from .configuration import Configuration

_LOGGER = logging.getLogger(__name__)

_OPENSHIFT = OpenShift()

_WORKFLOW_TEMPLATE_LABEL_SELECTOR = os.getenv(
    "THOTH_MOLDAVITE_TEMPLATE_LABEL_SELECTOR", "template=moldavite-jupyterbook-build"
)
_DEFAULT_BOOK_PATH = os.getenv("THOTH_MOLDAVITE_DEFAULT_BOOK_PATH", "book")
_DEFAULT_GIT_BRANCH = os.getenv("THOTH_MOLDAVITE_DEFAULT_GIT_BRANCH", "main")
_MAX_TTL = int(
    os.getenv("THOTH_MOLDAVITE_MAX_TTL", datetime.timedelta(days=3).total_seconds())
)
_DEFAULT_TTL = int(
    os.getenv("THOTH_MOLDAVITE_DEFAULT_TTL", datetime.timedelta(days=1).total_seconds())
)
_RESOURCES_CREATED = frozenset(
    {
        ("apps.openshift.io/v1", "DeploymentConfig"),
        ("argoproj.io/v1alpha1", "Workflow"),
        ("image.openshift.io/v1", "ImageStream"),
        ("route.openshift.io/v1", "Route"),
        ("v1", "Service"),
    }
)


def _get_openshift_resource(
    openshift: OpenShift,
    api_version: str,
    kind: str,
    label_selector: str,
    namespace: str,
) -> Dict[str, Any]:
    """Get a single pod name from a job."""
    # Kubernetes automatically adds 'job-name' label -> reuse it.
    response = openshift.ocp_client.resources.get(
        api_version=api_version, kind=kind
    ).get(
        namespace=namespace,
        label_selector=label_selector,
    )
    response = response.to_dict()
    if len(response["items"]) != 1:
        if len(response["items"]) > 1:
            # Log this error and report back to user not found.
            _LOGGER.error(
                "Multiple resources of kind %r (API version %r) for label selector %r found in %r",
                kind,
                api_version,
                label_selector,
                namespace,
            )

        raise NotFoundExceptionError(
            f"Multiple resources of kind {kind!r} (API version {api_version!r}) "
            f"for label selector {label_selector!r} found in {namespace!r}",
        )

    ret: Dict[str, Any] = response["items"][0]
    return ret


def get_version() -> Dict[str, Any]:
    """Obtain version identifier."""
    from moldavite import __version__
    from moldavite.entrypoint import __service_version__

    return {
        "version": __version__,
        "service_version": __service_version__,
    }


def post_jupyterbook_build(specification: Dict[str, Any]) -> Tuple[Dict[str, str], int]:
    """Create a JupyterBook build request."""
    repo_url = specification["repo_url"].strip()
    book_path = specification.get("book_path", _DEFAULT_BOOK_PATH)
    git_branch = specification.get("git_branch", _DEFAULT_GIT_BRANCH)
    ttl = int(specification.get("ttl", _DEFAULT_TTL))

    if ttl > _MAX_TTL:
        return {"error": f"TTL exceeded, maximum TTL can be {_MAX_TTL}"}, 400

    book_id = _OPENSHIFT.generate_id("book")

    _OPENSHIFT.workflow_manager.submit_workflow_from_template(
        namespace=Configuration.INFRA_NAMESPACE,
        label_selector=_WORKFLOW_TEMPLATE_LABEL_SELECTOR,
        template_parameters={
            "MOLDAVITE_REPO_URL": repo_url,
            "MOLDAVITE_BOOK_ID": book_id,
            "MOLDAVITE_REPO_BRANCH": git_branch,
            "MOLDAVITE_BOOK_PATH": book_path,
            "MOLDAVITE_TTL": ttl,
        },
        workflow_parameters=_OPENSHIFT._assign_workflow_parameters_for_ceph(),  # XXX: remove private call
        workflow_namespace=Configuration.BUILD_NAMESPACE,
    )

    return (
        {
            "book_id": book_id,
            "repo_url": repo_url,
            "git_branch": git_branch,
            "book_path": book_path,
        },
        202,
    )


def get_jupyterbook_status(book_id: str) -> Tuple[Dict[str, Any], int]:
    """Get status of the given JupyterBook."""
    # Workflow information.
    build_status = None
    try:
        build_status = _OPENSHIFT.get_workflow_status_report(
            workflow_id=book_id, namespace=Configuration.BUILD_NAMESPACE
        )
    except NotFoundExceptionError:
        _LOGGER.warning(
            f"No workflow {book_id!r} found in {Configuration.BUILD_NAMESPACE}"
        )

    # Information from Route.
    host = None
    try:
        route = _get_openshift_resource(
            _OPENSHIFT,
            api_version="route.openshift.io/v1",
            kind="Route",
            label_selector=f"build_id={book_id}",
            namespace=Configuration.BUILD_NAMESPACE,
        )
    except NotFoundExceptionError:
        _LOGGER.warning(
            f"No route {book_id!r} found in {Configuration.BUILD_NAMESPACE}"
        )
    else:
        host = route["status"]["ingress"][0]["host"]

    # Information from DeploymentConfig
    git_branch, book_path, repo_url, ttl = None, None, None, None
    try:
        deployment_config = _get_openshift_resource(
            _OPENSHIFT,
            api_version="apps.openshift.io/v1",
            kind="DeploymentConfig",
            label_selector=f"build_id={book_id}",
            namespace=Configuration.BUILD_NAMESPACE,
        )
    except NotFoundExceptionError:
        _LOGGER.warning(
            f"No DeploymentConfig {book_id!r} found in {Configuration.BUILD_NAMESPACE}"
        )
    else:
        git_branch = deployment_config["spec"]["template"]["metadata"]["annotations"][
            "moldavite.repo_branch"
        ]
        book_path = deployment_config["spec"]["template"]["metadata"]["annotations"][
            "moldavite.book_path"
        ]
        repo_url = deployment_config["spec"]["template"]["metadata"]["annotations"][
            "moldavite.repo_url"
        ]

        ttl_configured = int(
            deployment_config["metadata"]["labels"]["ttl"][:-1]
        )  # strip [s]
        creation = datetime_parser(deployment_config["metadata"]["creationTimestamp"])
        ttl = max(
            int(
                (
                    (creation + datetime.timedelta(seconds=ttl_configured))
                    - datetime.datetime.now(datetime.timezone.utc)
                ).total_seconds()
            ),
            0,
        )

    if build_status is None and ttl is None:
        return (
            {
                "error": f"Book {book_id!r} does not exist or the build has not started yet"
            },
            404,
        )

    return (
        {
            "host": host,
            "ttl": ttl,
            "book_id": book_id,
            "book_path": book_path,
            "git_branch": git_branch,
            "repo_url": repo_url,
            "build_status": build_status,
        },
        200,
    )


def delete_jupyterbook(book_id: str) -> Tuple[Dict[str, str], int]:
    """Delete the given JupyterBook."""
    any_found = False
    for api_version, kind in _RESOURCES_CREATED:
        response = _OPENSHIFT.ocp_client.resources.get(
            api_version=api_version,
            kind=kind,
        ).delete(
            namespace=Configuration.BUILD_NAMESPACE,
            label_selector=f"build_id={book_id}",
        )
        response = response.to_dict()
        if len(response["items"]) != 0:
            any_found = True

    if not any_found:
        return (
            {
                "error": f"Book {book_id!r} does not exist or the build has not started yet"
            },
            404,
        )

    return {"book_id": book_id}, 201


def get_jupyterbook_log(
    book_id: str,
) -> Tuple[Dict[str, Union[str, Dict[str, str]]], int]:
    """Get logs of the given JupyterBook workflow."""
    log = {}

    # XXX: eventually logs from other containers to help with issues debugging (e.g. wrong Git branch)
    try:
        log["build-book"] = _OPENSHIFT.get_workflow_node_log(
            node_name="build-book",
            workflow_id=book_id,
            namespace=Configuration.BUILD_NAMESPACE,
        )
    except NotFoundExceptionError:
        return {"error": f"No book with id {book_id!r} found"}, 404

    return {"log": log}, 200


def post_jupyterhub_build(specification: Dict[str, Any]) -> Tuple[Dict[str, str], int]:
    """Create a JupyterHub build request."""
    repo_url = specification["repo_url"].strip()
    git_branch = specification.get("git_branch", _DEFAULT_GIT_BRANCH)
    notebook_id = _OPENSHIFT.generate_id("notebook")

    _OPENSHIFT.workflow_manager.submit_workflow_from_template(
        namespace=Configuration.INFRA_NAMESPACE,
        label_selector="template=moldavite-jupyterhub-notebook-build",
        template_parameters={
            "MOLDAVITE_REPO_URL": repo_url,
            "MOLDAVITE_NOTEBOOK_ID": notebook_id,
            "MOLDAVITE_REPO_BRANCH": git_branch,
        },
        workflow_parameters=_OPENSHIFT._assign_workflow_parameters_for_ceph(),  # XXX: remove private call
        workflow_namespace=Configuration.BUILD_NAMESPACE,
    )

    return (
        {"notebook_id": notebook_id, "repo_url": repo_url, "git_branch": git_branch},
        202,
    )


def get_jupyterhub_status(notebook_id: str) -> Tuple[Dict[str, Any], int]:
    """Get status of the given JupyterHub NoteBook."""
    # Workflow information.
    build_status = None
    try:
        build_status = _OPENSHIFT.get_workflow_status_report(
            workflow_id=notebook_id, namespace=Configuration.BUILD_NAMESPACE
        )
    except NotFoundExceptionError:
        _LOGGER.warning(
            f"No workflow {notebook_id!r} found in {Configuration.BUILD_NAMESPACE}"
        )

    if build_status is None:
        return (
            {
                "error": f"NoteBook {notebook_id!r} does not exist or the build has not started yet"
            },
            404,
        )

    return (
        {
            "notebook_id": notebook_id,
            "build_status": build_status,
        },
        200,
    )


def delete_jupyterhub(notebook_id: str) -> Tuple[Dict[str, str], int]:
    """Delete the given JupyterHub NoteBook."""
    any_found = False
    for api_version, kind in _RESOURCES_CREATED:
        response = _OPENSHIFT.ocp_client.resources.get(
            api_version=api_version,
            kind=kind,
        ).delete(
            namespace=Configuration.BUILD_NAMESPACE,
            label_selector=f"build_id={notebook_id}",
        )
        response = response.to_dict()
        if len(response["items"]) != 0:
            any_found = True

    if not any_found:
        return (
            {
                "error": f"NoteBook {notebook_id!r} does not exist or the build has not started yet"
            },
            404,
        )

    return {"notebook_id": notebook_id}, 201


def get_jupyterhub_log(
    notebook_id: str,
) -> Tuple[Dict[str, Union[str, Dict[str, str]]], int]:
    """Get logs of the given JupyterHub NoteBook workflow."""
    log = {}

    # XXX: eventually logs from other containers to help with issues debugging (e.g. wrong Git branch)
    try:
        log["build-book"] = _OPENSHIFT.get_workflow_node_log(
            node_name="build-book",
            workflow_id=notebook_id,
            namespace=Configuration.BUILD_NAMESPACE,
        )
    except NotFoundExceptionError:
        return {"error": f"No notebook with id {notebook_id!r} found"}, 404

    return {"log": log}, 200
